idn/punycode.go

```diff
@@ -52,6 +59,12 @@ func ToPunycode(s string) string {
 
 // FromPunycode returns unicode domain name from provided punycode string.
 func FromPunycode(s string) string {
+	// Early check to see if decoding is needed.
+	// This will prevent making heap allocations when not needed.
+	if !strings.Contains(s, _PREFIX) {
```

Comment:
It's not very wrong. It will do exactly the same thing as it does now for those cases.

The NextLabel solution would be more correct but I guess also quite a bit slower. So the question is if you want to slow down the general case just to handle those special cases a bit better.